File: typescript/2022/04.ts
// https://adventofcode.com/2022/day/4

function parseLine(line: string): [number, number, number, number] {
  const [_, a, b, c, d] = line.match(/(\d+)-(\d+),(\d+)-(\d+)/)!;

  return [parseInt(a, 10), parseInt(b, 10), parseInt(c, 10), parseInt(d, 10)];
}

function part1(input: string): number {
  return input.split("\n").filter(line => {
    const [start1, end1, start2, end2] = parseLine(line);

    return (start1 <= start2 && end1 >= end2) || (start2 <= start1 && end2 >= end1);
  }).length;
}

function part2(input: string): number {
  return input.split("\n").filter(line => {
    const [start1, end1, start2, end2] = parseLine(line);

    return (start1 <= start2 && end1 >= start2) || (start2 <= start1 && end2 >= start1);
  }).length;
}

const input = (await Bun.file("../_input/2022/04.txt").text()).trimEnd();

if (import.meta.env.NODE_ENV === "test") {
  const { test, expect } = await import("bun:test");

  test("part 1", () => expect(part1(input)).toBe(584));
  test("part 2", () => expect(part2(input)).toBe(933));
} else {
  console.log("part 1:", part1(input));
  console.log("part 2:", part2(input));
}

File: rust/2022/src/bin/04.rs
// https://adventofcode.com/2022/day/4

use itertools::Itertools;
use regex::Regex;

fn parse_line(re: &Regex, line: &str) -> (u16, u16, u16, u16) {
    re.captures(line)
        .unwrap()
        .iter()
        .skip(1)
        .map(|m| m.unwrap().as_str().parse::<u16>().unwrap())
        .collect_tuple()
        .unwrap()
}

fn part1(input: &str) -> u16 {
    let re = Regex::new(r"(\d+)-(\d+),(\d+)-(\d+)").unwrap();

    input
        .lines()
        .filter(|line| {
            let (start1, end1, start2, end2) = parse_line(&re, line);
            (start1 <= start2 && end1 >= end2) || (start2 <= start1 && end2 >= end1)
        })
        .collect::<Vec<&str>>()
        .len() as u16
}

fn part2(input: &str) -> u16 {
    let re = Regex::new(r"(\d+)-(\d+),(\d+)-(\d+)").unwrap();

    input
        .lines()
        .filter(|line| {
            let (start1, end1, start2, end2) = parse_line(&re, line);
            (start1 <= start2 && end1 >= start2) || (start2 <= start1 && end2 >= start1)
        })
        .collect::<Vec<&str>>()
        .len() as u16
}

fn main() {
    let input = include_str!("../../../../_input/2022/04.txt").trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2: {}", part2(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../../../_input/2022/04.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 584);
    }

    #[test]
    fn part2() {
        assert_eq!(super::part2(INPUT.trim_end()), 933);
    }
}
